File: Go/Approach_2_O(n)_Time_&_O(n)_Space.go

// O(n) time | O(n) space

package program

func GetNthFib(n int) int {
	return helper(n, map[int]int{
		1: 0,
		2: 1,
	})
}

func helper(n int, memoize map[int]int) int {
	if val, found := memoize[n]; found {
		return val
	}
	memoize[n] = helper(n-1, memoize) + helper(n-2, memoize)
	return memoize[n]
}

File: Swift/Approach_2_O(n)_Time_&_O(n)_Space.swift

​// O(n) time | O(n) space

func getNthFib(n: Int) -> Int {
    var memoize = [1: 0, 2: 1]
    return helper(n: n, memoize: &memoize)
}

func helper(n: Int, memoize: inout [Int: Int]) -> Int {
    if memoize.keys.contains(n) {
        return memoize[n]!
    }
    memoize[n] = helper(n: n - 1, memoize: &memoize) + helper(n: n - 2, memoize: &memoize)
    return memoize[n]!
}
